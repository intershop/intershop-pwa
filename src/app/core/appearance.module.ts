import { NgModule } from '@angular/core';
import { NgbPopoverConfig } from '@ng-bootstrap/ng-bootstrap';
import { ToastrModule } from 'ngx-toastr';

import { IconModule } from './icon.module';
import { ThemeService } from './utils/theme/theme.service';

@NgModule({
  imports: [
    IconModule,
    ToastrModule.forRoot({
      closeButton: true,
      timeOut: 3000,
      positionClass: 'toast-top-full-width', // toast-top-center
      preventDuplicates: true,
    }),
  ],
})
export class AppearanceModule {
  constructor(popoverConfig: NgbPopoverConfig, themeService: ThemeService) {
    popoverConfig.placement = 'top';
    popoverConfig.triggers = 'hover';
    popoverConfig.container = 'body';

    themeService.init();
  }
}
