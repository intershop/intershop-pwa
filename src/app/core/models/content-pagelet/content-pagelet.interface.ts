import { ContentConfigurationParameterData } from 'ish-core/models/content-configuration-parameter/content-configuration-parameter.interface';
import { ContentSlotData } from 'ish-core/models/content-slot/content-slot.interface';

export interface ContentPageletData {
  definitionQualifiedName: string;
  id: string;
  displayName: string;
  domain: string;
  configurationParameters?: { [name: string]: ContentConfigurationParameterData };
  slots?: { [definitionQualifiedName: string]: ContentSlotData };
}
