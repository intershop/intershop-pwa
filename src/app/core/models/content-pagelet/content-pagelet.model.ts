import { ContentConfigurationParameters } from 'ish-core/models/content-configuration-parameter/content-configuration-parameter.mapper';
import { ContentSlot } from 'ish-core/models/content-slot/content-slot.model';

export interface ContentPagelet {
  definitionQualifiedName: string;
  id: string;
  displayName: string;
  domain: string;
  configurationParameters?: ContentConfigurationParameters;
  slots?: ContentSlot[];
}
