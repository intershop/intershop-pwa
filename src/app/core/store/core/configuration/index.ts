// API to access ngrx configuration state
export * from './configuration.actions';
export * from './configuration.selectors';
