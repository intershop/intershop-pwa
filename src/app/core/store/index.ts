export * from './actions';
export * from './effects';
export * from './reducers';
export * from './selectors';
export * from './serializers/router.serializers';
