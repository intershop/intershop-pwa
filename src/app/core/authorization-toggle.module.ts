import { ModuleWithProviders, NgModule } from '@angular/core';
import { ReplaySubject } from 'rxjs';
import { map } from 'rxjs/operators';

import { AuthorizationToggleDirective } from './directives/authorization-toggle.directive';
import { AuthorizationToggleService, checkPermission } from './utils/authorization-toggle/authorization-toggle.service';
import { whenTruthy } from './utils/operators';

@NgModule({
  declarations: [AuthorizationToggleDirective],
  exports: [AuthorizationToggleDirective],
})
export class AuthorizationToggleModule {
  private static permissions = new ReplaySubject<string[]>(1);

  static forTesting(...permissions: string[]): ModuleWithProviders<AuthorizationToggleModule> {
    AuthorizationToggleModule.switchTestingPermissions(...permissions);
    return {
      ngModule: AuthorizationToggleModule,
      providers: [
        {
          provide: AuthorizationToggleService,
          useValue: {
            isAuthorizedTo: (permission: string) =>
              AuthorizationToggleModule.permissions.pipe(
                whenTruthy(),
                map(perms => checkPermission(perms, permission))
              ),
          },
        },
      ],
    };
  }

  static switchTestingPermissions(...permissions: string[]) {
    AuthorizationToggleModule.permissions.next(permissions);
  }
}

export { AuthorizationToggleService } from './utils/authorization-toggle/authorization-toggle.service';
export { AuthorizationToggleGuard } from './guards/authorization-toggle.guard';
