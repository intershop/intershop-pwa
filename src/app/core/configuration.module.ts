import { NgModule } from '@angular/core';

import { SPECIAL_HTTP_ERROR_HANDLER } from './interceptors/icm-error-mapper.interceptor';
import { createPaymentErrorHandler } from './utils/http-error/create-payment.error-handler';
import { editPasswordErrorHandler } from './utils/http-error/edit-password.error-handler';
import { LoginUserErrorHandler } from './utils/http-error/login-user.error-handler';
import { requestReminderErrorHandler } from './utils/http-error/request-reminder.error-handler';
import { updatePasswordErrorHandler } from './utils/http-error/update-password.error-handler';

@NgModule({
  providers: [
    { provide: SPECIAL_HTTP_ERROR_HANDLER, useValue: updatePasswordErrorHandler, multi: true },
    { provide: SPECIAL_HTTP_ERROR_HANDLER, useClass: LoginUserErrorHandler, multi: true },
    { provide: SPECIAL_HTTP_ERROR_HANDLER, useValue: requestReminderErrorHandler, multi: true },
    { provide: SPECIAL_HTTP_ERROR_HANDLER, useValue: editPasswordErrorHandler, multi: true },
    { provide: SPECIAL_HTTP_ERROR_HANDLER, useValue: createPaymentErrorHandler, multi: true },
  ],
})
export class ConfigurationModule {}
