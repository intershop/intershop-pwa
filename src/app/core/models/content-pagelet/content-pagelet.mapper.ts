import { Injectable } from '@angular/core';

import { ContentConfigurationParameterMapper } from 'ish-core/models/content-configuration-parameter/content-configuration-parameter.mapper';
import { ContentSlotData } from 'ish-core/models/content-slot/content-slot.interface';
import { ContentSlot } from 'ish-core/models/content-slot/content-slot.model';

import { ContentPageletData } from './content-pagelet.interface';
import { ContentPagelet } from './content-pagelet.model';

@Injectable({ providedIn: 'root' })
export class ContentPageletMapper {
  constructor(private contentConfigurationParameterMapper: ContentConfigurationParameterMapper) {}

  fromData(data: ContentPageletData): ContentPagelet[] {
    if (!data) {
      throw new Error('falsy input');
    }

    const { definitionQualifiedName, id, domain, displayName } = data;
    const configurationParameters = this.contentConfigurationParameterMapper.fromData(data.configurationParameters);

    let slots: ContentSlot[] = [];
    let pagelets: ContentPagelet[] = [];
    if (data.slots) {
      const deep = Object.values(data.slots).map(x => this.fromSlotData(x));
      slots = deep.map(val => val.slot);
      pagelets = deep.map(val => val.pagelets).reduce((acc, val) => [...acc, ...val]);
    }

    return [
      {
        configurationParameters,
        definitionQualifiedName,
        id,
        slots,
        domain,
        displayName,
      },
      ...pagelets,
    ];
  }

  private fromSlotData(data: ContentSlotData): { slot: ContentSlot; pagelets: ContentPagelet[] } {
    if (!data) {
      throw new Error('falsy input');
    }

    const definitionQualifiedName = data.definitionQualifiedName;
    const configurationParameters = this.contentConfigurationParameterMapper.fromData(data.configurationParameters);
    const pageletIDs = !data.pagelets ? [] : data.pagelets.map(pagelet => pagelet.id);
    const displayName = data.displayName;

    const slot: ContentSlot = {
      definitionQualifiedName,
      configurationParameters,
      pageletIDs,
      displayName,
    };

    const pagelets = !data.pagelets
      ? []
      : data.pagelets.map(x => this.fromData(x)).reduce((acc, val) => [...acc, ...val], []);

    return { slot, pagelets };
  }
}
