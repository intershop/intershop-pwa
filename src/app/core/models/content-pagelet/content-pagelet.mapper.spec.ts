import { TestBed } from '@angular/core/testing';
import { provideMockStore } from '@ngrx/store/testing';

import { getICMStaticURL } from 'ish-core/store/core/configuration';

import { ContentPageletData } from './content-pagelet.interface';
import { ContentPageletMapper } from './content-pagelet.mapper';

describe('Content Pagelet Mapper', () => {
  let contentPageletMapper: ContentPageletMapper;

  beforeEach(() => {
    TestBed.configureTestingModule({
      providers: [
        provideMockStore({
          selectors: [{ selector: getICMStaticURL, value: 'http://www.example.org/static/channel/-' }],
        }),
      ],
    });

    contentPageletMapper = TestBed.inject(ContentPageletMapper);
  });

  it('should throw on empty input', () => {
    expect(() => contentPageletMapper.fromData(undefined)).toThrowError('falsy input');
  });

  it('should convert simple pagelet to single array instance', () => {
    const input: ContentPageletData = {
      id: 'pagelet',
      displayName: 'pagelet',
      domain: 'domain',
      definitionQualifiedName: 'domain-pagelet',
      configurationParameters: {
        key: {
          definitionQualifiedName: 'quali',
          value: 'test',
        },
      },
    };

    const result = contentPageletMapper.fromData(input);
    expect(result).toMatchSnapshot();
  });

  it('should convert pagelet with slots to single array instance', () => {
    const input: ContentPageletData = {
      id: 'pagelet',
      displayName: 'pagelet',
      domain: 'domain',
      definitionQualifiedName: 'domain-pagelet',
      configurationParameters: {
        key: {
          definitionQualifiedName: 'quali',
          value: 'test',
        },
      },
      slots: {
        slot1: {
          displayName: 'Slot',
          definitionQualifiedName: 'quali-slot',
          pagelets: [],
        },
      },
    };

    const result = contentPageletMapper.fromData(input);
    expect(result).toMatchSnapshot();
  });

  it('should convert pagelet with slots and additional pagelets to multi array instance', () => {
    const input: ContentPageletData = {
      id: 'pagelet',
      displayName: 'pagelet',
      domain: 'domain',
      definitionQualifiedName: 'domain-pagelet',
      configurationParameters: {
        key: {
          definitionQualifiedName: 'quali',
          value: 'test',
        },
      },
      slots: {
        slot1: {
          displayName: 'slot1',
          definitionQualifiedName: 'quali-slot',
          pagelets: [
            {
              id: 'pagelet-nested',
              definitionQualifiedName: 'fq',
              displayName: 'name-nested',
              domain: 'domain',
            },
          ],
        },
        slot2: {
          definitionQualifiedName: 'quali-slot',
          displayName: 'slot2',
          pagelets: [
            {
              id: 'pagelet-nested2',
              definitionQualifiedName: 'domain-pagelet',
              displayName: 'name1',
              domain: 'domain',
              configurationParameters: {
                key1: {
                  definitionQualifiedName: 'fq',
                  value: 'test-key1',
                },
              },
              slots: {
                slot11: {
                  definitionQualifiedName: 'fq-2',
                  displayName: 'slot11',
                  pagelets: [
                    {
                      id: 'pagelet-deeply-nested',
                      definitionQualifiedName: 'fq',
                      displayName: 'name-nested',
                      domain: 'domain',
                      configurationParameters: {
                        key3: {
                          definitionQualifiedName: 'fq',
                          value: '1',
                        },
                      },
                    },
                  ],
                },
              },
            },
          ],
          configurationParameters: {
            key1: {
              definitionQualifiedName: 'name1',
              value: 'hallo',
            },
            key2: {
              definitionQualifiedName: 'name2',
              value: ['hallo', 'welt'],
            },
          },
        },
      },
    };

    const result = contentPageletMapper.fromData(input);

    expect(result).toHaveLength(4);
    expect(result.map(p => p.id)).toIncludeAllMembers([
      'pagelet',
      'pagelet-nested',
      'pagelet-nested2',
      'pagelet-deeply-nested',
    ]);
    expect(result).toMatchSnapshot();
  });

  it('should have special handling for image pagelet configuration parmeters', () => {
    const input = {
      definitionQualifiedName: 'app_sf_base_cm:component.common.image.pagelet2-Component',
      displayName: 'Brand Image 5',
      domain: 'domain',
      id: 'cmp_brandImage_5',
      configurationParameters: {
        Image: {
          value: 'inSPIRED-inTRONICS-b2c-responsive:/brands/adata.jpg',
          definitionQualifiedName: 'app_sf_base_cm:component.common.image.pagelet2-Component-Image',
        },
      },
    } as ContentPageletData;

    expect(contentPageletMapper.fromData(input)[0]).toHaveProperty(
      'configurationParameters.Image',
      'http://www.example.org/static/channel/-/inSPIRED-inTRONICS-b2c-responsive/-/brands/adata.jpg'
    );
  });
});
