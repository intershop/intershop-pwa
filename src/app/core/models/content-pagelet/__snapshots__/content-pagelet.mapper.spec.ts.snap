// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Content Pagelet Mapper should convert pagelet with slots and additional pagelets to multi array instance 1`] = `
Array [
  Object {
    "configurationParameters": Object {
      "key": "test",
    },
    "definitionQualifiedName": "domain-pagelet",
    "displayName": "pagelet",
    "domain": "domain",
    "id": "pagelet",
    "slots": Array [
      Object {
        "configurationParameters": Object {},
        "definitionQualifiedName": "quali-slot",
        "displayName": "slot1",
        "pageletIDs": Array [
          "pagelet-nested",
        ],
      },
      Object {
        "configurationParameters": Object {
          "key1": "hallo",
          "key2": Array [
            "hallo",
            "welt",
          ],
        },
        "definitionQualifiedName": "quali-slot",
        "displayName": "slot2",
        "pageletIDs": Array [
          "pagelet-nested2",
        ],
      },
    ],
  },
  Object {
    "configurationParameters": Object {},
    "definitionQualifiedName": "fq",
    "displayName": "name-nested",
    "domain": "domain",
    "id": "pagelet-nested",
    "slots": Array [],
  },
  Object {
    "configurationParameters": Object {
      "key1": "test-key1",
    },
    "definitionQualifiedName": "domain-pagelet",
    "displayName": "name1",
    "domain": "domain",
    "id": "pagelet-nested2",
    "slots": Array [
      Object {
        "configurationParameters": Object {},
        "definitionQualifiedName": "fq-2",
        "displayName": "slot11",
        "pageletIDs": Array [
          "pagelet-deeply-nested",
        ],
      },
    ],
  },
  Object {
    "configurationParameters": Object {
      "key3": "1",
    },
    "definitionQualifiedName": "fq",
    "displayName": "name-nested",
    "domain": "domain",
    "id": "pagelet-deeply-nested",
    "slots": Array [],
  },
]
`;

exports[`Content Pagelet Mapper should convert pagelet with slots to single array instance 1`] = `
Array [
  Object {
    "configurationParameters": Object {
      "key": "test",
    },
    "definitionQualifiedName": "domain-pagelet",
    "displayName": "pagelet",
    "domain": "domain",
    "id": "pagelet",
    "slots": Array [
      Object {
        "configurationParameters": Object {},
        "definitionQualifiedName": "quali-slot",
        "displayName": "Slot",
        "pageletIDs": Array [],
      },
    ],
  },
]
`;

exports[`Content Pagelet Mapper should convert simple pagelet to single array instance 1`] = `
Array [
  Object {
    "configurationParameters": Object {
      "key": "test",
    },
    "definitionQualifiedName": "domain-pagelet",
    "displayName": "pagelet",
    "domain": "domain",
    "id": "pagelet",
    "slots": Array [],
  },
]
`;
